File: user_service/user_service.py
import grpc
from concurrent import futures
import user_pb2
import user_pb2_grpc
from models import User, SessionLocal
from datetime import datetime
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class UserService(user_pb2_grpc.UserServiceServicer):
    
    def CreateUser(self, request, context):
        db = SessionLocal()
        try:
            # Check if email already exists
            existing_user = db.query(User).filter(User.email == request.email).first()
            if existing_user:
                return user_pb2.CreateUserResponse(
                    success=False,
                    message="Email already exists"
                )
            
            # Create new user
            user = User(
                name=request.name,
                email=request.email,
                age=request.age
            )
            db.add(user)
            db.commit()
            db.refresh(user)
            
            logger.info(f"Created user: {user.id}")
            
            return user_pb2.CreateUserResponse(
                user=user_pb2.User(
                    id=user.id,
                    name=user.name,
                    email=user.email,
                    age=user.age,
                    created_at=user.created_at.isoformat()
                ),
                success=True,
                message="User created successfully"
            )
            
        except Exception as e:
            logger.error(f"Error creating user: {str(e)}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"Internal error: {str(e)}")
            return user_pb2.CreateUserResponse(success=False, message="Internal error")
        finally:
            db.close()
    
    def GetUser(self, request, context):
        db = SessionLocal()
        try:
            user = db.query(User).filter(User.id == request.user_id).first()
            
            if not user:
                return user_pb2.GetUserResponse(found=False)
            
            return user_pb2.GetUserResponse(
                user=user_pb2.User(
                    id=user.id,
                    name=user.name,
                    email=user.email,
                    age=user.age,
                    created_at=user.created_at.isoformat()
                ),
                found=True
            )
            
        except Exception as e:
            logger.error(f"Error getting user: {str(e)}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"Internal error: {str(e)}")
            return user_pb2.GetUserResponse(found=False)
        finally:
            db.close()
    
    def UpdateUser(self, request, context):
        db = SessionLocal()
        try:
            user = db.query(User).filter(User.id == request.user_id).first()
            
            if not user:
                context.set_code(grpc.StatusCode.NOT_FOUND)
                context.set_details("User not found")
                return user_pb2.UpdateUserResponse(
                    success=False,
                    message="User not found"
                )
            
            # Update user fields
            user.name = request.name
            user.email = request.email
            user.age = request.age
            
            db.commit()
            db.refresh(user)
            
            logger.info(f"Updated user: {user.id}")
            
            return user_pb2.UpdateUserResponse(
                user=user_pb2.User(
                    id=user.id,
                    name=user.name,
                    email=user.email,
                    age=user.age,
                    created_at=user.created_at.isoformat()
                ),
                success=True,
                message="User updated successfully"
            )
            
        except Exception as e:
            logger.error(f"Error updating user: {str(e)}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"Internal error: {str(e)}")
            return user_pb2.UpdateUserResponse(success=False, message="Internal error")
        finally:
            db.close()
    
    def DeleteUser(self, request, context):
        db = SessionLocal()
        try:
            user = db.query(User).filter(User.id == request.user_id).first()
            
            if not user:
                context.set_code(grpc.StatusCode.NOT_FOUND)
                context.set_details("User not found")
                return user_pb2.DeleteUserResponse(
                    success=False,
                    message="User not found"
                )
            
            db.delete(user)
            db.commit()
            
            logger.info(f"Deleted user: {request.user_id}")
            
            return user_pb2.DeleteUserResponse(
                success=True,
                message="User deleted successfully"
            )
            
        except Exception as e:
            logger.error(f"Error deleting user: {str(e)}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"Internal error: {str(e)}")
            return user_pb2.DeleteUserResponse(success=False, message="Internal error")
        finally:
            db.close()
    
    def ListUsers(self, request, context):
        db = SessionLocal()
        try:
            page = max(1, request.page or 1)
            limit = min(100, max(1, request.limit or 10))
            offset = (page - 1) * limit
            
            users = db.query(User).offset(offset).limit(limit).all()
            total = db.query(User).count()
            
            user_list = [
                user_pb2.User(
                    id=user.id,
                    name=user.name,
                    email=user.email,
                    age=user.age,
                    created_at=user.created_at.isoformat()
                )
                for user in users
            ]
            
            return user_pb2.ListUsersResponse(
                users=user_list,
                total=total,
                page=page,
                limit=limit
            )
            
        except Exception as e:
            logger.error(f"Error listing users: {str(e)}")
            context.set_code(grpc.StatusCode.INTERNAL)
            context.set_details(f"Internal error: {str(e)}")
            return user_pb2.ListUsersResponse(users=[], total=0, page=1, limit=10)
        finally:
            db.close()

def serve():
    port = "50051"
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    user_pb2_grpc.add_UserServiceServicer_to_server(UserService(), server)
    
    listen_addr = f"[::]:{port}"
    server.add_insecure_port(listen_addr)
    
    logger.info(f"Starting User Service on {listen_addr}")
    server.start()
    
    try:
        server.wait_for_termination()
    except KeyboardInterrupt:
        logger.info("Shutting down User Service")
        server.stop(0)

if __name__ == "__main__":
    serve()

File: user-service/src/user/user.service.ts
import { Injectable, Logger } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository } from "typeorm";
import { User } from "@/database/user.entity";
import {
  CreateUserRequest,
  CreateUserResponse,
  GetUserRequest,
  GetUserResponse,
  UpdateUserRequest,
  UpdateUserResponse,
  DeleteUserRequest,
  DeleteUserResponse,
  ListUsersRequest,
  ListUsersResponse,
} from "@/proto/user.pb";

@Injectable()
export class UserService {
  private readonly logger = new Logger(UserService.name);

  constructor(
    @InjectRepository(User)
    private readonly userRepository: Repository<User>
  ) {}

  async createUser(request: CreateUserRequest): Promise<CreateUserResponse> {
    try {
      // Check if email already exists
      const existingUser = await this.userRepository.findOne({
        where: { email: request.email },
      });

      if (existingUser) {
        return {
          user: undefined,
          success: false,
          message: "Email already exists",
        };
      }

      // Create new user
      const user = this.userRepository.create({
        name: request.name,
        email: request.email,
        age: request.age,
      });

      const savedUser = await this.userRepository.save(user);
      this.logger.log(`Created user: ${savedUser.id}`);

      return {
        user: {
          id: savedUser.id,
          name: savedUser.name,
          email: savedUser.email,
          age: savedUser.age,
          createdAt: savedUser.createdAt.toISOString(),
        },
        success: true,
        message: "User created successfully",
      };
    } catch (error) {
      this.logger.error(`Error creating user: ${error.message}`);
      return {
        user: undefined,
        success: false,
        message: "Internal error",
      };
    }
  }

  async getUser(request: GetUserRequest): Promise<GetUserResponse> {
    try {
      const user = await this.userRepository.findOne({
        where: { id: request.userId },
      });

      if (!user) {
        return {
          user: undefined,
          found: false,
        };
      }

      return {
        user: {
          id: user.id,
          name: user.name,
          email: user.email,
          age: user.age,
          createdAt: user.createdAt.toISOString(),
        },
        found: true,
      };
    } catch (error) {
      this.logger.error(`Error getting user: ${error.message}`);
      return {
        user: undefined,
        found: false,
      };
    }
  }

  async updateUser(request: UpdateUserRequest): Promise<UpdateUserResponse> {
    try {
      const user = await this.userRepository.findOne({
        where: { id: request.userId },
      });

      if (!user) {
        return {
          user: undefined,
          success: false,
          message: "User not found",
        };
      }

      // Update user fields
      user.name = request.name;
      user.email = request.email;
      user.age = request.age;

      const savedUser = await this.userRepository.save(user);
      this.logger.log(`Updated user: ${savedUser.id}`);

      return {
        user: {
          id: savedUser.id,
          name: savedUser.name,
          email: savedUser.email,
          age: savedUser.age,
          createdAt: savedUser.createdAt.toISOString(),
        },
        success: true,
        message: "User updated successfully",
      };
    } catch (error) {
      this.logger.error(`Error updating user: ${error.message}`);
      return {
        user: undefined,
        success: false,
        message: "Internal error",
      };
    }
  }

  async deleteUser(request: DeleteUserRequest): Promise<DeleteUserResponse> {
    try {
      const user = await this.userRepository.findOne({
        where: { id: request.userId },
      });

      if (!user) {
        return {
          success: false,
          message: "User not found",
        };
      }

      await this.userRepository.remove(user);
      this.logger.log(`Deleted user: ${request.userId}`);

      return {
        success: true,
        message: "User deleted successfully",
      };
    } catch (error) {
      this.logger.error(`Error deleting user: ${error.message}`);
      return {
        success: false,
        message: "Internal error",
      };
    }
  }

  async listUsers(request: ListUsersRequest): Promise<ListUsersResponse> {
    try {
      const page = Math.max(1, request.page || 1);
      const limit = Math.min(100, Math.max(1, request.limit || 10));
      const skip = (page - 1) * limit;

      const [users, total] = await this.userRepository.findAndCount({
        skip,
        take: limit,
        order: { createdAt: "DESC" },
      });

      const userList = users.map((user) => ({
        id: user.id,
        name: user.name,
        email: user.email,
        age: user.age,
        createdAt: user.createdAt.toISOString(),
      }));

      return {
        users: userList,
        total,
        page,
        limit,
      };
    } catch (error) {
      this.logger.error(`Error listing users: ${error.message}`);
      return {
        users: [],
        total: 0,
        page: 1,
        limit: 10,
      };
    }
  }
}
